cms-sw/cmssw
# implement a tracking particle associator with TrackCandidate and add trackCandidate branches in trackingNtuple

- SimDataFormats/Associations: 
    - add alias template to simplify definition of `SimToRecoCollection*` and `RecoToSimCollection*` and removed its partial duplicate definition in TrackToTrackingParticleAssociator.h to keep everything just in TrackAssociation.h. The fully defined types should be the same, so no update should be necessary in the `classes*` files for the dictionaries. There are no use cases at this point for placing `SimToRecoCollectionTCandidate` and `RecoToSimCollectionTCandidate` in `edm::Event`; so, no dictionary is added for this.
    - changed `TrackToTrackingParticleAssociator::associateRecoToSim` arguments/interface from `Handle<TrackingParticleCollection` to `RefVector<TrackingParticleCollection` to be more in common with existing use cases for `reco::Tracks`
- SimTracker/TrackAssociatorProducers: improve templated methods in `QuickTrackAssociatorByHitsImpl` so that `reco::Track` and `TrackCandidate` have maximal overlap in implementation and minimize copy-pasted code
- Validation/RecoTrack: trackingNtuple updates
    - add `tcand_*` branches to provide access to `TrackCandidate` details (optional, enabled for extendedContent custom option). This should allow more straightforward debugging of track pattern recognition, in particular in mkFit. This data can also be used to train track candidate classification and allow removing candidates from the following final fit step (and save computation)
    - refactor mergeIters case (where built tracks are used as inputs) to provide access to per-iteration MVAs
    - add (global) hit positions for rechits as computed with the track-trajectory corrected CPEs to allow for hit position and residual analysis
    - cleanup the outdated use of `TransientTrackingRecHitBuilder` and use tracking hits directly. This is in a sense a somewhat late follow-up to #3344 

No changes are expected in the standard workflows.
If that is confirmed, there may be a backport to an older release (12_2_X where the TRK-Run3Winter22 are made, in case it's not practical to run on those inputs in 12_4_X)
